### PR TITLE
workflow_state: reject invalid configurations

### LINKED ISSUE #6157: workflow-state: handle submitted as a message/trigger not status
The workflow-state command defaults to polling task status rather than task trigger.

The CLI upgrade advise tells you to upgrade an existing message poller into a status poller:

```console
$ cylc workflow-state generic --point 20191209T1200Z --task bar --message=submitted
WARNING - --task, --status, --message, --output, --point, --task-point are deprecated. Please use an ID:
    generic//20191209T1200Z/bar:submitted
```

This is inconsistent with the xtrigger upgrade advice which tells you to use `is_message=True`:

```
        workflow_state(message=submitted, point=1234, task=foo, workflow=ab) -->
    workflow_state(workflow_task_id=ab//1234/foo:submitted, is_message=True)
```

The CLI advice becomes a problem in the case of the `submitted` output. Because `submitted` is a status, it defaults to status polling. However, the submitted status is transitory so this test is not safe.

Suggest:
* Changing the CLI upgrade advice to match the xtrigger advice.
* Raising a warning if workflow status polling is configured for non-final statuses.
* Document trigger polling as the preferred approach.

### PR DESCRIPTION
**Pull request by oliver-sanders**
* Do not allow users to poll for transient statuses.
* Reject invalid task states.
* Reject polling waiting and preparing tasks (not reliably pollable).
* Closes #6157

(minor bug as upgrade advice encouraged users to adopt invalid configurations which were not subsequently rejected)

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

File: tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc [hunk 1]
```diff
@@ -8,7 +8,7 @@
     initial cycle point = 2011
     final cycle point = 2016
     [[xtriggers]]
-        upstream = workflow_state("{{UPSTREAM}}//%(point)s/foo:data_ready"):PT1S
+        upstream = workflow_state("{{UPSTREAM}}//%(point)s/foo:data_ready", is_trigger=True):PT1S
```
Comment (oliver-sanders) [inline tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc:11]: Behaviour change.

Turned off the implicit, "this isn't a status so handle it as a trigger" behaviour for xtriggers to make them explicit for safety.

File: cylc/flow/scripts/workflow_state.py [hunk 2]
```diff
@@ -427,6 +429,15 @@ def main(parser: COP, options: 'Values', *ids: str) -> None:
...
         else:
             msg += id_.replace(options.depr_point, "$CYLC_TASK_CYCLE_POINT")
+
+        if (
+            options.depr_status
+            and options.depr_status in TASK_STATUSES_ALL
+            and options.depr_status not in TASK_STATUSES_FINAL
+        ):
+            # polling for non-final task stauses is flaky
+            msg += 'and the --triggers option'
```
Comment (MetRonnie) [inline cylc/flow/scripts/workflow_state.py:439]:
```suggestion
            msg += ' and the --triggers option'
```

Comment (MetRonnie) [reply cylc/flow/scripts/workflow_state.py:439]: Although arguably this is unnecessary as it is adequately covered by the `InputError`

**Comment by MetRonnie** [comment]
Ack, just noticed this regression:

Cylc 8.2.7:
```
$ cylc workflow-state Murphid -t foo -p 2001 -S catacomb --max-polls 1
InputError: invalid status 'catacomb'
```

Cylc 8.3.1 and this branch:
```
$ cylc workflow-state Murphid -t foo -p 2001 -S catacomb --max-polls 1
WARNING - --task, --status, --message, --output, --point, --task-point are deprecated. Please use an ID: Murphid//2001/foo:catacomb
ERROR - failed after 1 polls
```

**Comment by oliver-sanders** [comment]
> Ack, just noticed this regression:

Just with the deprecated `-S` presumably, easy fix.

File: cylc/flow/scripts/workflow_state.py [hunk 3]
```diff
@@ -427,6 +429,15 @@ def main(parser: COP, options: 'Values', *ids: str) -> None:
             msg += id_
         else:
             msg += id_.replace(options.depr_point, "$CYLC_TASK_CYCLE_POINT")
+
+        if (
+            options.depr_status
+            and options.depr_status in TASK_STATUSES_ALL
+            and options.depr_status not in TASK_STATUSES_FINAL
+        ):
+            # polling for non-final task stauses is flaky
```
Comment (hjoliver) [inline cylc/flow/scripts/workflow_state.py:438]:
```suggestion
            # polling for non-final task statuses is flaky
```

**Review by hjoliver — APPROVED** [review]
My bad, but need `--trigger` and `--message` --> `--triggers` and `--messages` here: https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26|

**Review by hjoliver — APPROVED** [review]
https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26

My bad: need to pluralize `--trigger` and `--message` here

**Comment by oliver-sanders** [comment]
> My bad, but need --trigger and --message --> --triggers and --messages here: [https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26|](https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26%7C)

Link appears to be duff, where was is pointing?

File: changes.d/fix.6175.md
Comment (MetRonnie) [inline changes.d/fix.6175.md:1]: Should be `6175.fix.md`

(Using `towncrier create` prevents this mistake)

File: tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc (hunk 1 above)
Comment (MetRonnie) [reply tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc:11]: Why do this for the xtrigger but not the CLI?

**Comment by oliver-sanders** [comment]
> Why do this for the xtrigger but not the CLI?

Back compatibility.

I would prefer the CLI to work this way, and maybe we could make that change, but it would be an embarrassing reversal between 8.3.0 and 8.3.3.

WDYT?

**Comment by MetRonnie** [comment]
Would it be worth having a deprecation warning until 8.4 and then making it an InputError (either for the CLI or both CLI and xtrigger)? Although I don't think it looks great, not sure it's worse than having the xtrigger already an InputError and nothing for the CLI.

**Comment by oliver-sanders** [comment]
If Hillary is ok with this I'm more happy to change the CLI behaviour.

Note that xtriggers are validated as part of workflow startup, whereas the CLI command is not validated until it is run, so it is easier to push xtrigger interface changes than.

**Comment by hjoliver** [comment]
> > My bad, but need --trigger and --message --> --triggers and --messages here: [https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26|](https://github.com/cylc/cylc-flow/pull/6175/files#diff-16e1252035a0b4a9605f8cf4f45f5d3c4df8d90923b480137d43166beb734113R25-R26%7C)
> 
> Link appears to be duff, where was is pointing?

Branch probably got rebased. ...

**Comment by hjoliver** [comment]
> If Hillary is ok with this I'm more happy to change the CLI behaviour.

I'm happy with that.

Note that I considered this during the main workflow-state overhaul PR, but decided against it because (a) historically we defaulted to polling for status; and (b) the command and xtrigger names also suggest that.  But if y'all are happy with the change, let's make it more robust.

File: changes.d/fix.6175.md
Comment (oliver-sanders) [reply changes.d/fix.6175.md:1]: Must have followed the pattern of the [other entries](https://github.com/cylc/cylc-flow/blob/d7540aa62ed8a67f32f21508b7143ef10c03c13c/changes.d/fix.5924.md).

(`towncrier create` is too much faff)

File: tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc (hunk 1 above)
Comment (oliver-sanders) [reply tests/flakyfunctional/xtriggers/01-workflow_state/flow.cylc:11]: The same reason this feature was originally implemented, back-compat.

But we now seem to be ok with dropping this (which is much nicer).

File: changes.d/fix.6175.md
Comment (MetRonnie) [reply changes.d/fix.6175.md:1]: In towncrier v24.7 there is now an interactive mode if you run `towncrier create` without any args, seems like even less effort than manually creating a fragment file to me

**Comment by MetRonnie** [comment]
(Test failures)

**Review by MetRonnie — APPROVED** [review]

**Pull request merged**